File: picasso-transformations/src/main/java/io/github/hendraanggrian/picassotransformations/color/ColorOverlayTransformation.java
package io.github.hendraanggrian.picassotransformations.color;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;

import io.github.hendraanggrian.picassotransformations.BaseTransformation;
import io.github.hendraanggrian.picassotransformations.Key;
import io.github.hendraanggrian.picassotransformations.internal.PaintBuilder;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public class ColorOverlayTransformation extends BaseTransformation {

    public static final String TAG = "overlay";
    public static final String NAME_COLOR = "color";

    private int color;

    public ColorOverlayTransformation(@ColorInt int color) {
        this.color = color;
    }

    @NonNull
    @Override
    public Bitmap transform(@NonNull Bitmap source, int height, int width, @NonNull Bitmap target) {
        new Canvas(target).drawBitmap(source, 0, 0, new PaintBuilder(Paint.ANTI_ALIAS_FLAG)
                .colorFilter(new PorterDuffColorFilter(color, PorterDuff.Mode.SRC_ATOP))
                .build());
        source.recycle();
        return target;
    }

    @NonNull
    @Override
    public Key getKey() {
        return Key.fromTag(TAG).put(NAME_COLOR, color);
    }
}

File: picasso-utils/src/com/hendraanggrian/picasso/transformation/ColorOverlayTransformation.kt
package com.hendraanggrian.picasso.transformation

import android.graphics.*
import android.support.annotation.ColorInt
import com.squareup.picasso.Transformation

internal class ColorOverlayTransformation(@ColorInt private val color: Int) : Transformation {

    override fun transform(source: Bitmap): Bitmap {
        val target = Bitmap.createBitmap(source.width, source.height, Bitmap.Config.ARGB_8888)
        Canvas(target).drawBitmap(source, 0f, 0f, Paint(Paint.ANTI_ALIAS_FLAG).apply {
            colorFilter = PorterDuffColorFilter(color, PorterDuff.Mode.SRC_ATOP)
        })
        source.recycle()
        return target
    }

    override fun key() = "ColorOverlayTransformation[color=$color]"
}